### DISCUSSION TITLE
[FEATURE REQUEST] Register vectorized python UDFs for use in SQL statements

### DISCUSSION THREAD

**Discussion by NickCrews** [duckdb/duckdb#4797] [Ideas]
Currently, [you can create UDFs in python](https://github.com/duckdb/duckdb/pull/1569) and use them using `duckdb.from_df(df).map(my_udf)`.

I would like to be able to register my UDF with duckdb such that I can call it from within SQL statements, similar to how you can register pandas DataFrames, eg. `con.register('my_table', df)`. This is also similar to how you can register custom python UDFs for use in pyspark.

I propose something like:

```python
# external python library
import nicknames

nn = nicknames.NickNamer()

def _is_nickname_of(n1: str, n2: str) -> bool:
    return n1 in nn.nicknames_of(n2)

def is_nickname_of(name1: pd.Series, name2: pd.Series) -> pd.Series:
    # This is sort of a silly example because I'm just python iterating over
    # the values, but in a real use case I would have a library that takes and returns
    # pyarrow arrays, and then this could actually be fast.
    result = (_is_nickname_of(n1, n2) for n1, n2 in zip(name1, name2))
    return pd.Series(result, index=name1.index)

con.register('is_nickname_of', is_nickname_of)
con.execute("SELECT is_nickname_of('Rob', 'Robert')").fetchone()
```
The motivation for this is I am trying to use [splink](https://github.com/moj-analytical-services/splink) for fuzzy deduplication, and in that framework you [specify comparison functions using SQL statements](https://moj-analytical-services.github.io/splink/topic_guides/customising_comparisons.html#method-3-providing-the-spec-as-a-dictionary). I want to be able to compare using `is_nickname_of(first_name_l, first_name_r) OR is_nickname_of(first_name_r, first_name_l)` I can do this currently using splink's pyspark backend, but it would be great to be able to do it with duckDB as well so I could swap between backends for exploration (duckdb) and full runs (pyspark)

**Comment by hannes** [comment]
What you describe here are so-called scalar UDFs. They are called "Scalar" because they - at least conceptionally - apply to a single row of data in a single call and also return a single Scalar value as a result.

We have thought about **Scalar** Python UDFs at length already. Implementing those not immediately straightforward due to two main issues: 

1. **Function Call Overhead**: Calling a function in Python is rather expensive. It will be slow to loop over rows of data calling the function for every single row/value. 
2. **Global Interpreter Lock**: We cannot call into the Python interpreter from multiple threads and get a good speedup because of Python's GIL. Hence we cannot directly use parallelism to alleviate issue 1 a bit.

There are some ideas that we could implement Python UDFs 1 and 2, they each come with their own drawbacks. 
- **UDF Servers**: This is what APACHE Spark does for example. Start a couple of separate Python processes, ship the input data to those processes, run the UDFs there, and ship the results back. This is tricky, DuckDB would need to use sys calls or fork to create more Python processes, manage them somewhat (what happens if one dies etc), reload any Python packages in them, deal with serialisation of data, contention etc. The mechanics of an in-process database doing this are questionable to say the least.
- **JIT-Vectorized UDFs**: We could use something like Numba to generate a native function (Just-In-Time compilation) that wraps the user-supplied function to run on a larger chunk of data at the same time, much like DuckDB's internal execution engine works. The advantage here is that we do not need a separate process, we would need to study if the Numba semantics exactly match the Python ones, and for which kind of functions the JIT would work for. Not so sure either.

There are some good news, we have already implemented table-valued UDFs for Python. These functions get a chunk of the input table passed as a Pandas data frame, can manipulate it in whatever way they want and are required to return a data frame again. Again, the UDF only sees a chunk of the input at a time. Here is an example

```Python
import duckdb
conn = duckdb.connect()
query_result = conn.query("SELECT * FROM generate_series(10)")
print(query_result)
map_result = query_result.map(lambda df: df['generate_series'].add(42).to_frame())
print(map_result)
```

Here, the table-valued UDF is the lambda in the call to `.map`, it just adds 42 to the series in the query result.

**Reply by NickCrews** [reply]
Thank you for the great explanation. I had a vague understanding of the challenges of ser/deser, but I didn't know many of these details!

Sorry, my example was a bit misleading because my vector -> vector UDF then immediately goes and uses a scalar -> scalar helper function. Perhaps it was also confusing because it was annotated as pd.Series -> pd.Series, when perhaps it should have been pd.DataFrame -> pd.DataFrame.

I'm looking for exactly what is already implemented (pd.DataFrame -> pd.DataFrame), but just accessible from the SQL API. In the library I'm trying to use, I can't use the python `my_df.map(my_func)` syntax, I have to supply SQL commands.

At first glance of how [map was implemented](https://github.com/duckdb/duckdb/pull/1569/files#diff-0fc07f8b0433fe421126a7adaffede4508e7d8a5f4cb65e209539b44ea22ef9fR21-R42), it looks like a lot of this logic already is happening in the C++ layer, so I imagined it could work as

1. when you call `con.register('is_nickname_of', is_nickname_of)`, then at the C++ layer you add an entry to a lookup table of type `string` -> `PyObject *`, where the PyObject is the opaque handle to the python function.
2. Later, in a SQL query, when you come across a function call to "is_nickname_of", then you can reference this lookup table, find the stored python function, and then reuse the existing logic. This is maybe similar to how macro lookup happens?

Does that clear things up?

**Reply by hannes** [reply]
Yes we can totally make this work so you can use the function from SQL.

**Reply by NickCrews** [reply]
@hannes thank you, that would be excellent! I can help contribute at a mid level if you point me in the right direction. Maybe reopen https://github.com/duckdb/duckdb/issues/4795 so we can track progress there?

**Comment by tobymao** [comment]
@hannes I think you're overthinking scalar UDF's and trying to make them performant. I would also really like scalar UDF's in duckdb and I don't care about performance, I just need them to work. So I hope it wouldn't be difficult to implement something simple to use and register.

My use case is mainly using Duckdb for CI / Unit testing / Compatibility. I think scalar UDF's would be valuable regardless of performance. Having them available but slow would be a great value add imo.

**Reply by hannes** [reply]
Thing is, ca. 5 mins after we add the slow ones there is another discussion thread in this forum about them being slow. But I would be happy to discuss this further. I propose joins the DuckDB Foundation so we can have a more structured discussion about such roadmap issues.

**Reply by tobymao** [reply]
Where is the DuckDB Foundation?

**Reply by Mytherin** [reply]
You can find information about the [DuckDB foundation here](https://duckdb.org/foundation/)

**Reply by tobymao** [reply]
Ah, I no longer work at a company that can afford to be in the DuckDB foundation.

**Comment by tobymao** [comment]
hey @hannes is there any update on this feature request?

**Reply by hannes** [reply]
No.

**Comment by Tishj** [comment]
I've got a branch [here](https://github.com/tishj/duckdb/tree/python_table_function) that can do this:
```py
import duckdb

from duckdb.typing import *

con = duckdb.connect()

def table_func(df):
	print(df)
	return df

con.register_table_function('no_op_tablefunc', table_func)
rel2 = con.sql('select * from no_op_tablefunc((from range(10)))')
print(rel2)
```
```
Empty DataFrame
Columns: [range]
Index: []
Empty DataFrame
Columns: [range]
Index: []
   range
0      0
1      1
2      2
3      3
4      4
5      5
6      6
7      7
8      8
9      9
┌───────┐
│ range │
│ int64 │
├───────┤
│     0 │
│     1 │
│     2 │
│     3 │
│     4 │
│     5 │
│     6 │
│     7 │
│     8 │
│     9 │
└───────┘
```

This just utilizes the same code as `map`, only instead of using it on a relation you can register as a function to be used in SQL queries.
Note that like Hannes already mentioned, this is not optimized in any way yet

**Reply by NickCrews** [reply]
Brilliant! Am I interpreting that in this case `df` is a pyarrow Table? I would prefer that over a pandas DF, it is more universal.

**Reply by Tishj** [reply]
No not currently, it uses `map` which works with pandas dataframes

**Comment by tekumara** [comment]
Scalar UDFs have been implemented in #7171

**Comment by pohuan** [comment]
https://github.com/pohuan/duckdb/pull/3
I'm working on user defined aggregation function in python for school project.
Currently I'm replacing combine/finalize functions with python. It works. But is there any better interface that allow user to replace only one function?

**Comment by Goykhman** [comment]
To poke at this thread, also in the context of @hannes mentioning [JIT-vectorized UDFs](https://github.com/duckdb/duckdb/discussions/4797#discussioncomment-3726080), we have been working with @nelson2005 on the [NumbDuck](https://github.com/Goykhman/numbduck)  library that dynamically taps into the DuckDB C API from Python and exposes it via [numba-JIT'd](https://numba.readthedocs.io/en/stable/user/jit.html) proxies.

The NumbDuck project could generally be of use for Python applications that need to avoid Python interpreter overhead and accelerate the computationally heavy bits using [numba](https://numba.pydata.org/), while still being able to call DuckDB from within JIT-compiled functions.

But specifically in the context of this thread's discussion, we have added JIT'd [scalar](https://github.com/Goykhman/numbduck/blob/main/numbduck/ducklib.py#L1329-L1428) and [aggregate](https://github.com/Goykhman/numbduck/blob/main/numbduck/ducklib.py#L1431-L1512) proxies for the DuckDB C API for [scalar](https://duckdb.org/docs/current/clients/c/api#scalar-functions)  and [aggregate](https://duckdb.org/docs/current/clients/c/api#aggregate-functions) UDFs.

While scalar UDFs have been integrated into the DuckDB Python client, aggregate UDFs are currently supported via the C API but not exposed as a native Python API.

In NumbDuck, we have some [demos](https://github.com/Goykhman/numbduck/tree/main/examples) for both [scalar](https://github.com/Goykhman/numbduck/blob/main/examples/haversine.py) and [aggregate](https://github.com/Goykhman/numbduck/blob/main/examples/irr.py) user-defined functions, including performance benchmarks (where available) against the Python and PyArrow counterparts.

NumbDuck was also discussed [here](https://github.com/duckdb/duckdb-python/issues/404#issuecomment-4231135525) with @evertlammerts, albeit not in the context of UDFs.

**Comment by kinthaiofficial** [comment]
Vectorized Python UDFs in DuckDB would be a significant addition — the current row-at-a-time model creates a Python call overhead per row that completely negates DuckDB's vectorized execution advantages.

The gap: DuckDB internally processes data in columnar batches (vectors). When you call a Python UDF, you cross the C++→Python boundary per row, serializing/deserializing each value individually. For UDFs on large tables, this is often the bottleneck.

**What vectorized UDFs would enable**:
```python
# Current: row-at-a-time (slow)
duckdb.create_function("my_udf", lambda x: x * 2, [int], int)

# Proposed: vectorized (fast — processes entire column batch)
@duckdb.vectorized_function(input_types=[pa.int64()], return_type=pa.int64())
def my_udf(col: pa.Array) -> pa.Array:
    return pa.compute.multiply(col, 2)
```

The key is that the UDF receives an Arrow Array (or chunked array), does bulk computation (via numpy/Arrow compute), and returns an Arrow Array. This keeps data in the native vectorized representation throughout.

**Alternative today**: Arrow-based extensions via the C API. If you're doing complex transforms, implementing as a DuckDB C extension with Arrow integration gets you vectorized performance. But the Python dev experience is much worse.

**Related pattern**: for ML inference use cases (embedding generation, model scoring), this is especially valuable — running a model on a batch of 1024 rows via numpy is 100-1000x faster than running it row by row.

Is this for a specific use case (ML inference, string processing, custom aggregation)? The implementation complexity varies significantly depending on whether you need stateful or stateless UDFs.

**Reply by nelson2005** [reply]
Just to clarify... @Goykhman repo has both stateful and stateless python vectorized UDFs that run at C speed, no python interpreter involved.  There are plain-python, pyarrow, and fully-jitted examples, along with benchmarks.
Are you asking about functionality that's not demonstrated?  Or a simpler syntax?  Or am I misunderstanding completely, which I frequently do? :)

**Comment by rustyconover** [comment]
@kinthaiofficial I have this vectorized feature coming in an upcoming DuckDB extension, get in touch if you'd like a preview.

**Comment by rustyconover** [comment]
As a really simple example here is a scalar function that just multiplies

```python
class MultiplyFunction(ScalarFunction):
    """Multiplies a value by a constant factor.

    This example demonstrates type inference with array classes:
    - pa.Int64Array -> pa.int64() (inferred from Annotated type)
    - ConstParam() for constant scalar input (receives Python value at runtime)
    - Returns() output type is also inferred from pa.Int64Array

    Example:
        SQL:    SELECT multiply(price, 2) FROM products
        Input:  price=[10, 20, 30]
        Args:   factor=2
        Output: result=[20, 40, 60]

    """

    class Meta:
        """Function metadata."""

        name = "multiply"
        description = "Multiplies a value by a constant factor"
        examples = [
            FunctionExample(
                sql="SELECT multiply(price, 2) FROM products",
                description="Double all prices",
            ),
            FunctionExample(
                sql="SELECT multiply(quantity, 10) FROM inventory",
                description="Scale quantities by 10",
            ),
        ]

    @classmethod
    def compute(
        cls,
        value: Annotated[pa.Int64Array, Param(doc="Integer value to multiply")],
        factor: Annotated[int, ConstParam("Multiplication factor")],
    ) -> Annotated[pa.Int64Array, Returns()]:
        """Multiply values by the constant factor."""
        return pc.multiply(value, factor)
```
